driver/convert.c

```diff
@@ -1653,6 +1755,1012 @@ static SQLRETURN wstr_to_time(esodbc_rec_st *arec, esodbc_rec_st *irec,
+			wltrim_ws(wstr);
+			PARSE_IVL_FLD_OR_RET(rec, /*hour limit*/23, wstr,
+					&ivl->intval.day_second.hour,
+					/*end?*/ivl->interval_type == SQL_IS_DAY_TO_HOUR);
+			if (ivl->interval_type == SQL_IS_DAY_TO_HOUR) {
+				break;
+			}
+			if (wstr->str[0] != L':') {
+				ERRH(stmt, "invalid char as separator: `%c`.", wstr->str[0]);
+				RET_HDIAGS(stmt, SQL_STATE_22018);
+			} else {
+				wstr->str ++;
+				wstr->cnt --;
+			}
+			PARSE_IVL_FLD_OR_RET(rec, /*minute limit*/59, wstr,
+					&ivl->intval.day_second.minute,
+					/*end?*/ivl->interval_type == SQL_IS_DAY_TO_MINUTE);
+			if (ivl->interval_type == SQL_IS_DAY_TO_MINUTE) {
+				break;
+			}
+			if (wstr->str[0] != L':') {
+				ERRH(stmt, "invalid char as separator: `%c`.", wstr->str[0]);
+				RET_HDIAGS(stmt, SQL_STATE_22018);
+			} else {
+				wstr->str ++;
+				wstr->cnt --;
+			}
+			ret = parse_interval_second(rec, /*second limit*/59, wstr, ivl);
+			if (! SQL_SUCCEEDED(ret)) {
+				return ret;
+			}
+			break;
+
+		case SQL_IS_HOUR_TO_MINUTE:
+		case SQL_IS_HOUR_TO_SECOND:
+			PARSE_IVL_FLD_OR_RET(rec, /*limit*/0, wstr,
+					&ivl->intval.day_second.hour, /*end?*/FALSE);
+			if (wstr->str[0] != L':') {
+				ERRH(stmt, "invalid char as separator: `%c`.", wstr->str[0]);
+				RET_HDIAGS(stmt, SQL_STATE_22018);
+			} else {
+				wstr->str ++;
+				wstr->cnt --;
+			}
+			/*no break*/
+		case SQL_IS_MINUTE_TO_SECOND:
+			PARSE_IVL_FLD_OR_RET(rec,
+					/*minute limit*/
+					ivl->interval_type == SQL_IS_MINUTE_TO_SECOND ? 0 : 59,
+					wstr, &ivl->intval.day_second.minute,
+					/*end?*/ivl->interval_type == SQL_IS_HOUR_TO_MINUTE);
+			if (ivl->interval_type == SQL_IS_HOUR_TO_MINUTE) {
+				break;
+			}
+			if (wstr->str[0] != L':') {
+				ERRH(stmt, "invalid char as separator: `%c`.", wstr->str[0]);
+				RET_HDIAGS(stmt, SQL_STATE_22018);
+			} else {
+				wstr->str ++;
+				wstr->cnt --;
+			}
+			ret = parse_interval_second(rec, /*second limit*/59, wstr, ivl);
+			if (! SQL_SUCCEEDED(ret)) {
+				return ret;
+			}
+			break;
+	}
+	/*INDENT-ON*/
+
+	return SQL_SUCCESS;
+#	undef PARSE_IVL_FLD_OR_RET
+}
+
+/* parse the qualifier in `± '<value>' <qualifier>`, rtrim-ing it from wstr */
+static SQLSMALLINT parse_interval_type(wstr_st *wstr)
+{
+	/* compares the end of the local `wstr` against given `_end` string and
+	 * trims the former, on match.
+	 * Uses local var: wstr.
+	 * Returns on no match (the qualifier is incorrect). */
+#	define TRIM_IF_ENDS_WITH_OR_RET(_end) \
+	do { \
+		wstr_st _wend = MK_WSTR(_end); \
+		if (wstr->cnt <= _wend.cnt || \
+			wmemncasecmp(&wstr->str[wstr->cnt - _wend.cnt], \
+				_wend.str, _wend.cnt)) { \
+			return 0; \
+		} \
+		wstr->cnt -= _wend.cnt; \
+		wrtrim_ws(wstr); \
+	} while (0)
+#	define IS_DELIM(wc) \
+	(wc == L'\'' || wc == L'.' || (L'0' <= wc && wc <= L'9' ))
+
+
+	if (wstr->cnt <= /* "day", smallest interval qualifier */3) {
+		return 0;
+	}
+
+	/* split by last letter */
+	switch (wstr->str[wstr->cnt - 1] | 0x20) {
```

Comment:
I assume the `| 0x20` bits are to make the letters lower case?  Is this a standard C idiom?  I haven't seen it used widely, although I'm a C++ programmer so that might just be ignorance of idiomatic C.

If it's not a mega popular idiom then:

1. Are you sure it won't cause a subtle bug where a character that's not a letter gets converted to some other character that matches a `case`?
2. Even if 1 is not a problem, it might be worth adding a comment to say why the `| 0x20` is there.

Alternatively, could `towupper` be used?

Reply:
> I assume the | 0x20 bits are to make the letters lower case?

Correct. 

> Is this a standard C idiom?

For the ASCII charset, yes, that's common-ish (as a cheap `tolower()` alternative).  For a whole UTF8 set, this wouldn't work, but SQL keywords only uses the ASCII set, so this should be safe. 

>  a character that's not a letter gets converted to some other character that matches a case

This should not happen, since `| 0x20` can only increase character's value and the comparison is done against a "full" wide character with a value from within the ASCII subset (i.e. no truncation occurs, which could lead to a false match).

> it might be worth adding a comment to say why the | 0x20 is there.

Sure, I'll add it.